### ISSUE TITLE
Removal of getters in InterceptedInvocation breaks Mockito-Scala

### ISSUE THREAD

**Issue by bbonanno** [mockito/mockito#1987]
Hi there,

I was upgrading to mockito-core 3.4.x and noticed my build breaks, after looking at the commit history I can see the commit b3fc349e8e525e41f34ff04aca8c3140ff2b335a removed 3 getters from `InterceptedInvocation` that I rely on to work out some of the Scala magic, is there any specific reason why they were removed? if so, is there any alternative way to obtain access to that info? if not, can we re-instate them?

Thanks!

**Comment by TimvdLippe** [comment]
It seems like these methods were not part of our interfaces and thus we dont include them in our public API breakage policy. I suppose we could add them, if there are no alternative APIs that you could use. Could you check if you have any alternatives available? If not, we can reconsider promoting them to a public API for mockito-scala. Cc @raphw

**Comment by bbonanno** [comment]
Right now I have my own `MockHandler` that intercepts the invocations, and does some stuff, for it I'm currently matching on the type `InterceptedInvocation` given the fact that `Invocation` doesn't have all the info I need (these methods being a big part of it). 
I also have my own implementation of `Invocation`, to construct it I also need all the information available in the core impl.

It would be nicer to couple myself to `Invocation` and not `InterceptedInvocation` tbh, so promoting those fields to that interface (which I assume is the public API) would be great and would shield me a bit more of changes (or total replacement) on `InterceptedInvocation`

In case you wanna take a quick look, [this](https://github.com/mockito/mockito-scala/blob/release/1.x/common/src/main/scala/org/mockito/internal/handler/ScalaMockHandler.scala) is the `ScalaMockHandler` I've implemented, its main responsibilities are:

a) executing compiler generated methods that should never be mocked and are necessary for a consistent behaviour of the lang.
b) transforming varargs, dealing with implicit resolution and Equals matchers and a bunch of other type related things.
c) the decoration is also different as I need my own null guardian too, due to differences in the type system.

TL;DR I second the motion of adding those methods as part of the public API as it would decouple some parts of mockito-scala from internals of mockito-core

**Comment by raphw** [comment]
We can add the getters again, they are not used within Mockito but I'll mark them as deprecated. They are part of the internal API and should not be used by external dependencies.

**Comment by bbonanno** [comment]
mockito-scala is part of the mockito ecosystem, we need a solution that lasts in time...

**Comment by raphw** [comment]
Fully agree, it's just problematic for us to retain compatibility within our *internal* package. We marked it as such to avoid clashes like this one being reported. Ideally, we find a way to turn such usage into official API.

**Comment by bbonanno** [comment]
Yes, that's exactly why I was suggesting adding those methods to the public API (or maybe we can think in having the concept of an "Integration API" for libs that rely a bit more on the internals?)

**Comment by raphw** [comment]
You are right, this does however require exposing several additional interfaces that currently reside in the internal API. Those are `MockReference`, `MockitoMethod` and `RealMethod`.

Could you prepare a PR where you expose a minimal set of API in `Invocation` that you require to make mockito-scala work?

**Comment by bbonanno** [comment]
Sure, I'll raise a PR as soon as I can

**Comment by bbonanno** [comment]
I just took a look and the problem is that I need to construct my own invocation, so I need access to those raw values.
The good news is that my invocation lives in the same package, so we could make those methods package protected and that should do the trick right? or you don't wanna have any methods that don't exist in a public interface?

`getMockRef` and `getMockitoMethod` are only used so I can construct the new invocation
`getRealMethod` is used cause I need to call `isInvokable` as I need sometimes to actually call the method if I detect it's actually a synthetic method created by the compiler.

What doy think?

**Comment by raphw** [comment]
I don't think package equality is a good thing to aim for since the module system requires exclusive packages.

We should aim for a good, official API that remains stable. Worst case, you'd need to copy some of our internal code. All SPI is public so you have a chance to recreate what we already created without relying on our internals. If you'd like to reuse some of it, we can however try to factor it out.

**Comment by TimvdLippe** [comment]
This was fixed in #1665
